CMakeLists.txt

```diff
@@ -1,47 +1,109 @@
+endif()
+if(POLICY CMP0135)
+    cmake_policy(SET CMP0135 NEW)
+endif()
 if (${CMAKE_BUILD_TYPE} STREQUAL "Release")
     add_definitions (-DNOPAssert)
 endif ()
 
 add_compile_options (-Wall)
 add_compile_options (-Wunused)
 add_compile_options (-Wextra)
-add_compile_options (-Werror)
+#add_compile_options (-Werror)
 
 # allow deprecated functions
 add_compile_options (-Wno-deprecated-declarations)
 
 option (USE_STATIC_LIBRARIES "Link with static libraries if available" ON)
 
-message (STATUS "Build type is: " ${CMAKE_BUILD_TYPE})
+if(NOT IPPL_PLATFORMS)
+    message(STATUS "${Yellow}IPPL_PLATFORMS flag is not set, defaulting to serial backend${ColourReset}")
+    set(IPPL_PLATFORMS "SERIAL")
+endif()
 
-find_package(Kokkos 4.0.0 REQUIRED)
+if(IPPL_PLATFORMS)
+    #message(STATUS "IPPL_PLATFORMS flag is set to: ${IPPL_PLATFORMS}")
+    string(TOUPPER "${IPPL_PLATFORMS}" IPPL_PLATFORMS)
+    # Perform case distinction
+    if("${IPPL_PLATFORMS}" STREQUAL "CUDA")
+        message(STATUS "${Green}CUDA backend enabled${ColourReset}")
+        set(Kokkos_ENABLE_CUDA ON CACHE BOOL "Enable Kokkos CUDA Backend")
+        set(Heffte_ENABLE_CUDA ON CACHE BOOL "Set Heffte CUDA Backend")
+    elseif("${IPPL_PLATFORMS}" STREQUAL "OPENMP")
+        message(STATUS "${Green}OpenMP backend enabled${ColourReset}")
+        if(NOT Heffte_ENABLE_FFTW)
+            message(STATUS "${Yellow}Because Heffte_ENABLE_FFTW is not set defaulting to Stock with AVX${ColourReset}")
+            set(Heffte_ENABLE_AVX2 ON CACHE BOOL "Enable Heffte Stock backend lol")
+        else()
+            message(STATUS "${Green}Enabled FFTW${ColourReset}")
+        endif()
+        set(Kokkos_ENABLE_OPENMP ON CACHE BOOL "Enable Kokkos OpenMP Backend")
+    elseif("${IPPL_PLATFORMS}" STREQUAL "SERIAL")
+
+        message(STATUS "${Green}Serial backend enabled${ColourReset}")
+        if(NOT Heffte_ENABLE_FFTW)
+            message(STATUS "${Yellow}Because Heffte_ENABLE_FFTW is not set defaulting to Stock with AVX${ColourReset}")
+            set(Heffte_ENABLE_AVX2 ON CACHE BOOL "Enable Heffte Stock backend lol")
+        endif()
+
+    elseif("${IPPL_PLATFORMS}" STREQUAL "CUDA;OPENMP" OR "${IPPL_PLATFORMS}" STREQUAL "OPENMP;CUDA")
+        message(STATUS "${Green}CUDA and OpenMP backend enabled${ColourReset}")
+        set(Heffte_ENABLE_CUDA ON CACHE BOOL "Set Heffte CUDA Backend")
+        set(Kokkos_ENABLE_OPENMP ON CACHE BOOL "Enable Kokkos OpenMP Backend")
+        set(Kokkos_ENABLE_CUDA ON CACHE BOOL "Enable Kokkos CUDA Backend") 
+    else()
+        message(FATAL_ERROR "Unknown or unsupported platform: ${IPPL_PLATFORMS}")
+        # Handle unknown or unsupported platform
+    endif()
+endif()
 
+find_package(Kokkos 4.0.0 QUIET)
+if(NOT Kokkos_FOUND)
+    FetchContent_Declare(
+        #DOWNLOAD_EXTRACT_TIMESTAMP
+        kokko
+        URL https://github.com/kokkos/kokkos/archive/refs/tags/4.1.00.tar.gz                
```

Comment:
I see that this is hard-coded to fetch Kokkos version 4.1.00. Can't we make this more general?

Reply:
We can do that via variable:
```
set(KOKKOS_VERSION "4.1.00")

# Use the variable in FetchContent_Declare
FetchContent_Declare(
    kokkos
    URL https://github.com/kokkos/kokkos/archive/refs/tags/${KOKKOS_VERSION}.tar.gz
)
```
or more sophisticated via major and minor? Let me know your thoughts.

Reply:
I think passing the Kokkos version like this is fine. However, we must check if the file exists.

Reply:
I addressed it: specifying no kokkos version yields:
![image](https://github.com/IPPL-framework/ippl/assets/5041338/45db5770-8cc0-452d-9272-cafb1aee6353)
Otherwise: `cmake .. -DKOKKOS_VERSION=4.2.00` yields:
![image](https://github.com/IPPL-framework/ippl/assets/5041338/3c17656e-86da-4972-8b77-632b620c31cc)

---

For heffte:
Default behaviour (`cmake .. -DENABLE_FFT=True`) is again:
![image](https://github.com/IPPL-framework/ippl/assets/5041338/c945a1b1-6b95-48fd-8ae2-a948d01c94e4)
For a specified version `cmake .. -DENABLE_FFT=True -DHEFFTE_VERSION=2.4.0`:
![image](https://github.com/IPPL-framework/ippl/assets/5041338/937985c6-404d-4df1-b926-ea6e388e7fb0)

Reply:
Yes, that is fine. I am just wondering about wrong inputs. What happens if a release does not exist? What does `FetchContent_Declare` do in such a case?

Reply:
Currently it just prints a verbose "download failed" message

Reply:
Does cmake print this message and abort?

Reply:
Yes. To be precise, `FetchContent_MakeAvailable` attempts a https download and the github server responds with an error message `22`, for example:
```
    error: downloading 'https://github.com/kokkos/kokkos/archive/refs/tags/4.43.00.tar.gz' failed
          status_code: 22
          status_string: "HTTP response code said error"
          log:
          --- LOG BEGIN ---
          Host github.com:443 was resolved.
```

Reply:
OK. When cmake aborts, this is fine for me.